File: src/main/java/com/ramusthastudio/plugin/unixtimestamp/hints/BaseInlayHintsCollector.java
package com.ramusthastudio.plugin.unixtimestamp.hints;

import com.intellij.codeInsight.hints.FactoryInlayHintsCollector;
import com.intellij.codeInsight.hints.InlayHintsSink;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiElement;
import com.ramusthastudio.plugin.unixtimestamp.settings.AppSettingsState;
import com.ramusthastudio.plugin.unixtimestamp.utils.Helper;
import org.jetbrains.annotations.NotNull;

public final class BaseInlayHintsCollector<T extends PsiElement> extends FactoryInlayHintsCollector {
  private final AppSettingsState settingsState;
  private final Class<T> psiElement;

  public BaseInlayHintsCollector(@NotNull Editor editor,
      AppSettingsState settingsState,
      Class<T> psiElement) {
    super(editor);
    this.settingsState = settingsState;
    this.psiElement = psiElement;
  }

  @Override
  public boolean collect(@NotNull PsiElement element, @NotNull Editor editor, @NotNull InlayHintsSink sink) {
    if (psiElement.isInstance(element)) {
      Helper.createInlayHintsElement(element, sink, getFactory(), settingsState);
      return true;
    }
    return false;
  }
}

File: src/main/kotlin/com/ramusthastudio/plugin/unixtimestamp/hints/BaseInlayHintsCollector.kt
package com.ramusthastudio.plugin.unixtimestamp.hints

import com.intellij.codeInsight.hints.declarative.InlayTreeSink
import com.intellij.codeInsight.hints.declarative.InlineInlayPosition
import com.intellij.codeInsight.hints.declarative.HintFormat
import com.intellij.codeInsight.hints.declarative.OwnBypassCollector
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.progress.ProgressManager
import com.intellij.psi.PsiFile
import com.ramusthastudio.plugin.unixtimestamp.settings.AppSettingsState
import com.ramusthastudio.plugin.unixtimestamp.utils.Helper.createInstantFormat
import com.ramusthastudio.plugin.unixtimestamp.utils.Helper.findUnixTimestamp

class BaseInlayHintsCollector(
    private val editor: Editor,
    private val psiFileClass: Class<out PsiFile>,
    private val settingsState: AppSettingsState = AppSettingsState.instance
) : OwnBypassCollector {
    override fun collectHintsForFile(file: PsiFile, sink: InlayTreeSink) {
        ProgressManager.checkCanceled()
        val document = editor.document
        if (!psiFileClass.isInstance(file) || document.textLength > MAX_DOCUMENT_LENGTH) {
            return
        }

        findUnixTimestamp(
            document.charsSequence,
            settingsState.isSupportMicroSecondsEnable,
            settingsState.isSupportNanoSecondsEnable,
            MAX_HINTS_PER_FILE
        ).forEachIndexed { index, (word, textRange) ->
            if (index % CANCELLATION_CHECK_INTERVAL == 0) {
                ProgressManager.checkCanceled()
            }

            val instant = createInstantFormat(word) ?: return@forEachIndexed
            val hint = settingsState.defaultLocalFormatter.format(instant)
            val offset = if (settingsState.isInlayHintsPlaceEndOfLineEnable) {
                textRange.endOffset
            } else {
                textRange.startOffset
            }
            sink.addPresentation(
                InlineInlayPosition(offset, false),
                emptyList(),
                null,
                HintFormat.default
            ) { text(hint) }
        }
    }

    companion object {
        internal const val MAX_DOCUMENT_LENGTH = 2_000_000
        internal const val MAX_HINTS_PER_FILE = 1_000
        private const val CANCELLATION_CHECK_INTERVAL = 128
    }
}
